h2/src/main/org/h2/util/DateTimeUtils.java

```diff
@@ -1479,39 +1479,79 @@ public static long convertScale(long nanosOfDay, int scale) {
      * @return date truncated to 'day'
      */
     public static Value truncateDate(String timeUnit, Value value) {
-        Value result = null;
+        Value result;
 
-        // Retrieve the dateValue.
+        // Retrieve the dateValue and the time in nanoseconds if the date.
         long[] fieldDateAndTime = DateTimeUtils.dateAndTimeFromValue(value);
         long dateValue = fieldDateAndTime[0];
+        long timeNanosRetrieved = fieldDateAndTime[1];
 
-        // Case where the date has to be truncated to the day.
-        if (timeUnit.equals("DAY")) {
+        // Variable that contains the number of nanoseconds in 'hour', 'minute',
+        // etc.
+        long nanoInHour = 3_600_000_000_000l;
+        long nanoInMinute = 60_000_000_000l;
+        long nanoInSecond = 1_000_000_000l;
+        long nanoInMilliSecond = 1_000_000l;
+        long nanoInMicroSecond = 1_000l;
 
-            if (value instanceof ValueTimestampTimeZone) {
+        // Variable used to the time in nanoseconds of the date truncated.
+        long timeNanos;
 
-                // Create a new ValueTimestampTimeZone by only setting the
-                // date. The time in nanoseconds since midnight will be set
-                // to 0.
-                ValueTimestampTimeZone vTmp = (ValueTimestampTimeZone) value;
-                result = ValueTimestampTimeZone.fromDateValueAndNanos(vTmp.getDateValue(), 0,
-                        vTmp.getTimeZoneOffsetMins());
+        // Compute the number of time unit in the date, for example, the
+        // number of time unit 'HOUR' in '15:14:13' is '15'. Then convert the
+        // result to nanoseconds.
+        if (timeUnit.equals("MICROSECONDS")) {
```

Comment:
There are two possibilities.

1. Use `switch` over string here, Java can do it and such switch should be more readable.

2. Use `getDatePart()` in caller to get a numeric constant. Use switch with these constants here. This should be even better. In this case you will need to add a mappings for `DECADE` and `CENTURY` later to common mappings like it was done for `EPOCH`. I'm not sure about `DECADE`, but mapping for `CENTURY` may be useful in another similar functions. Of course, if you're planning to implement these cases, it's not necessary to support all possible values right now.